kong/plugins/acl/groups.lua

```diff
@@ -138,9 +124,60 @@ local function get_current_consumer_id()
 end
 
 
+--- Returns a table with all group names.
+-- The table will have an array part to iterate over, and a hash part
+-- where each group name is indexed by itself. Eg.
+-- {
+--   [1] = "users",
+--   [2] = "admins",
+--   users = "users",
+--   admins = "admins",
+-- }
+-- If there are no authenticated_groups defined, it will return nil
+-- @return table with groups or nil
+local function get_authenticated_groups()
+  local authenticated_groups = kong.ctx.shared.authenticated_groups
+  if type(authenticated_groups) ~= "table" then
+    authenticated_groups = ngx.ctx.authenticated_groups
+    if authenticated_groups == nil then
+      return nil
+    end
+
+    if type(authenticated_groups) ~= "table" then
+      kong.log.warn("invalid authenticated_groups, a table was expected")
+      return nil
+    end
+  end
+
+  local groups = {}
+  for i = 1, #authenticated_groups do
+    groups[i] = authenticated_groups[i]
+    groups[authenticated_groups[i]] = authenticated_groups[i]
+  end
+
+  return groups
+end
+
+
+--- checks whether a group-list is part of a given list of groups.
+-- @param groups_to_check (table) an array of group names.
+-- @param groups (table) list of groups (result from
+-- `get_authenticated_groups`)
+-- @return (boolean) whether the authenticated group is part of any of the
+-- groups.
+local function group_in_groups(groups_to_check, groups)
+  for i = 1, #groups_to_check do
+    if groups[groups_to_check[i]] then
```

Comment:
As far as I am concerned, this `groups.lua` liven in `kong.plugins.acl.groups`. I have hard time seeing anyone would include that in anywhere, so in that sense I consider the `groups.lua` internal only and if you where to use it elsewhere, you are probably going to hit yourself to the foot at some point. The `groups.lua` is very plugin depended. Maybe we can move that inside the `acl.lua` but I think I have moved it back and forth in this review. If we where to make groups something `core` functionality, this whole groups thing should be rethought and rewritten.

Reply:
`groups.lua` has been factored out in a separate file because it is also used by the canary plugin (enterprise). This was temporary such that the coupling would be on a single file which was to dissapear in the PDK somewhere in the future

Reply:
This text in the PR description:

> the support to this needs to be implemented in the plugins as well

seems to indicate that the `groups` variable is supposed to be shared with other plugins.

> groups.lua has been factored out in a separate file because it is also used by the canary plugin (enterprise)

I'm then quite hesitant to approve further changes to `groups.lua`. This is the kind of situation we introduced the PDK for.

Reply:
> > `groups.lua` has been factored out in a separate file because it is also used by the canary plugin (enterprise)
>
>I'm then quite hesitant to approve further changes to `groups.lua`. This is the kind of situation we introduced the PDK for.

more specifically: the canary plugin has a copy of the `groups.lua`. So the file can be changed, but what they do share is the same database structures. The groups for the canary plugin are configured through the acl plugin endpoints.

Reply:
We introduced PDK, but getting groups support to PDK is a totally different and separate talk to this PR. It may be that we don't want groups to be anything close to core feature, or we might, but then implement it totally differently. What this PR adds is `ngx.ctx.authenticated_groups` or `kong.ctx.shared.authenticated_groups` and that's it. Some plugins already implement support for it (though it does not work until this gets merged) while other may not, but may later. The only plugin to consume this context value is this `acl` plugin.

Why do we need this? Because with say `basic-auth` you can use ACL plugin, but without this PR merged you cannot use ACL plugin with say OpenID Connect plugin. I feel this PR held as a victim of endless conversation.

Reply:
All right, as clarified by @Tieske, this module is not really used outside of the plugin, so there's no problem in changing its interface. So I think the changes in `groups.lua` are ok.